cardano-db-sync/src/Cardano/DbSync/Cache.hs

```diff
@@ -361,6 +369,72 @@ queryPrevBlockWithCache msg cache hsh =
       liftIO $ missPrevBlock (cStats ci)
       liftLookupFail msg $ DB.queryBlockId hsh
 
+queryTxIdWithCache ::
+  MonadIO m =>
+  CacheStatus ->
+  Ledger.TxId StandardCrypto ->
+  ByteString ->
+  Text ->
+  ExceptT SyncNodeError (ReaderT SqlBackend m) DB.TxId
+queryTxIdWithCache cache ledgerTxId txHash errTxt = do
+  case cache of
+    NoCache -> liftLookupFail errTxt $ DB.queryTxId txHash
+    ActiveCache ci -> do
+      mp <- liftIO $ readTVarIO (cTxIds ci)
+      case FIFO.lookupTxIdCache ledgerTxId mp of
+        Just txId -> do
+          liftIO $ hitTxIds (cStats ci)
+          pure txId
+        Nothing -> do
+          resTxId <- liftLookupFail errTxt $ DB.queryTxId txHash
+          liftIO $ missTxIds (cStats ci)
+          liftIO $ atomically $ modifyTVar (cTxIds ci) $ FIFO.insertTxIdCache ledgerTxId resTxId
+          pure resTxId
+
+resolveInputTxId ::
+  MonadIO m =>
+  Generic.TxIn ->
+  CacheStatus ->
+  ReaderT SqlBackend m (Either DB.LookupFail DB.TxId)
+resolveInputTxId txIn cache = do
```

Comment:
There is a lot of duplication between this and `queryTxIdWithCache`

Reply:
i tried to fix this duplication but the problem is the call site happen at different locations when `TxIn` is and isn't accessible. I managed to abstract out the cache logic but the insertion logic required a bunch of  higher-order functions and it made it way more confusing to read imo